File: src/main/java/pw/mihou/amelia/commands/feeds/FeedsCommand.java
package pw.mihou.amelia.commands.feeds;

import org.javacord.api.entity.channel.ServerTextChannel;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.permission.Role;
import org.javacord.api.entity.server.Server;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.message.MessageCreateEvent;
import pw.mihou.amelia.commands.base.Command;
import pw.mihou.amelia.commands.db.FeedDB;
import pw.mihou.amelia.models.FeedModel;
import pw.mihou.amelia.models.FeedNavigator;
import pw.mihou.amelia.templates.Embed;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

public class FeedsCommand extends Command {

    public FeedsCommand() {
        super("feeds", "Returns back all the feeds on the server.", "feeds", false);
    }

    @Override
    protected void runCommand(MessageCreateEvent event, User user, Server server, String[] args) {
        FeedDB.getServer(server.getId()).getModels().thenAccept(feedModels -> {
            FeedNavigator navigator = new FeedNavigator(feedModels);
            if (!navigator.getModels().isEmpty()) {
                event.getMessage().reply(embed(server, navigator.current().orElse(new ArrayList<>()), 1)).thenAccept(message -> {
                    if (navigator.hasNext()) {
                        message.addReactions("⬅", "trash:775601666845573140", "➡");
                    } else {
                        message.addReaction("trash:775601666845573140");
                    }
                    message.addReactionAddListener(e -> {
                        if (e.getUserId() == event.getMessageAuthor().getId()) {
                            if (e.getEmoji().equalsEmoji("➡")) {
                                if (navigator.hasNext()) {
                                    message.edit(embed(server, navigator.next().orElse(new ArrayList<>()), navigator.getPage()));
                                }
                            } else if (e.getEmoji().equalsEmoji("⬅")) {
                                if (navigator.canReverse()) {
                                    message.edit(embed(server, navigator.backwards().orElse(new ArrayList<>()), navigator.getPage()));
                                }
                            }
                            if (e.getEmoji().getMentionTag().equalsIgnoreCase("<:trash:775601666845573140>")) {
                                message.delete();
                                event.getMessage().delete();
                                navigator.reset();
                            }
                        }

                        if (e.getUserId() != event.getApi().getYourself().getId()) {
                            e.removeReaction();
                        }
                    }).removeAfter(5, TimeUnit.MINUTES).addRemoveHandler(() -> {
                        message.removeAllReactions();
                        navigator.reset();
                    });
                });
            } else {
                event.getMessage().reply(embed(server, new ArrayList<>(), 1));
            }
        });
    }

    private EmbedBuilder embed(Server server, ArrayList<FeedModel> objects, int page) {
        EmbedBuilder embed = new Embed().setTitle(server.getName() + "'s feeds").setFooter("Page: " + page)
                .setDescription(!objects.isEmpty() ? "Here are the feeds registered on the server." : "The server has no feeds registered.").build();
        if (!objects.isEmpty()) {
            for (FeedModel object : objects) {
                StringBuilder builder = new StringBuilder();
                object.getMentions().forEach(aLong -> builder.append(server.getRoleById(aLong).map(Role::getMentionTag).orElse("[Unknown role]")));
                embed.addField("[" + object.getUnique() + "] " + object.getName(), "\n" +
                        "\nLink: " + object.getFeedURL() +
                        "\nFeed Unique ID: `" + object.getUnique() +
                        "`\nFeed ID: `" + object.getId() +
                        "`\nFeed Name: `" + object.getName() +
                        "`\nRoles Subscribed: " + builder.toString() +
                        "\nLast Update: `" + object.getDate().toString() +
                        "`\nAssigned Channel: " + server.getTextChannelById(object.getChannel())
                        .map(ServerTextChannel::getMentionTag).orElse("Unknown (possibly deleted?)") +
                        "\nCreated by: " + server.getMemberById(object.getUser()).map(User::getMentionTag).orElse("Unknown (possibly left?)"));
            }
        }
        return embed.setFooter("Please use the Unique ID for removing feeds, etc.");
    }

}

File: src/main/java/pw/mihou/amelia/commands/FeedsCommand.kt
package pw.mihou.amelia.commands

import org.javacord.api.entity.message.component.Button
import org.javacord.api.entity.message.embed.EmbedBuilder
import org.javacord.api.entity.server.Server
import org.javacord.api.entity.user.User
import org.javacord.api.event.interaction.ButtonClickEvent
import org.javacord.api.interaction.SlashCommandOption
import org.javacord.api.interaction.callback.InteractionOriginalResponseUpdater
import pw.mihou.amelia.db.FeedDatabase
import pw.mihou.amelia.models.FeedModel
import pw.mihou.amelia.scheduledExecutorService
import pw.mihou.amelia.utility.redactListLink
import pw.mihou.nexus.features.command.facade.NexusCommandEvent
import pw.mihou.nexus.features.command.facade.NexusHandler
import pw.mihou.nexus.features.command.validation.OptionValidation
import pw.mihou.nexus.features.command.validation.errors.ValidationError
import pw.mihou.nexus.features.paginator.NexusPaginatorBuilder
import pw.mihou.nexus.features.paginator.enums.NexusPaginatorButtonAssignment
import pw.mihou.nexus.features.paginator.facade.NexusPaginatorCursor
import pw.mihou.nexus.features.paginator.facade.NexusPaginatorEvents
import java.awt.Color
import java.util.concurrent.TimeUnit

@Suppress("UNUSED")
object FeedsCommand: NexusHandler {

    private const val name = "feeds"
    private const val description = "Views all the feeds that are registered for this server."

    private val options = listOf(
        SlashCommandOption.createLongOption("id", "The id of the feed to look-up.", false)
    )

    val validators = listOf(
        OptionValidation.create(
            collector = { event -> event.interaction.getArgumentLongValueByName("id") },
            validator = { id -> id <= 9999 },
            error = { ValidationError.create("A feed identifier has at maximum 4 digits of numbers (0-9999).") },
            requirements = OptionValidation.createRequirements {
                nonNull = createErrorableRequireSetting(ValidationError.Companion.create("You cannot leave the `feed` option."))
            }
        )
    )

    override fun onEvent(event: NexusCommandEvent) {
        val server = event.server.orElse(null)

        if (server == null) {
            event.respondNowAsEphemeral().setContent("❌ You cannot use this command in a private channel.").respond()
            return
        }

        when (event.interaction.getArgumentLongValueByName("id").isPresent) {
            true -> {
                val id = event.interaction.getArgumentLongValueByName("id").get()
                val feed = FeedDatabase.get(id)

                if (feed == null) {
                    event.respondNowAsEphemeral().setContent("❌ There are no feeds with the id of **$id**.").respond()
                    return
                }

                event.respondNowAsEphemeral().addEmbed(embed(server, feed, event.user)).respond()
            }
            false -> {
                val chunks = FeedDatabase.all(server.id).chunked(5)

                if (chunks.isEmpty()) {
                    event.respondNow().setContent("❌ There are no feeds registered on this server, please use the `/register` command to register one!").respond()
                    return
                }

                NexusPaginatorBuilder<List<FeedModel>>(chunks)
                    .setButton(NexusPaginatorButtonAssignment.NEXT, Button.secondary("", "Next"))
                    .setButton(NexusPaginatorButtonAssignment.PREVIOUS, Button.secondary("", "Previous"))
                    .setEventHandler(object: NexusPaginatorEvents<List<FeedModel>> {
                        override fun onInit(
                            updater: InteractionOriginalResponseUpdater,
                            cursor: NexusPaginatorCursor<List<FeedModel>>
                        ): InteractionOriginalResponseUpdater = updater.addEmbed(embed(server, cursor))

                        override fun onPageChange(cursor: NexusPaginatorCursor<List<FeedModel>>, event: ButtonClickEvent) {
                            event.buttonInteraction.message.edit(embed(server, cursor))
                        }
                    })
                    .build()
                    .send(event.baseEvent.interaction, event.respondLater().join())
                    .thenAccept {  instance ->
                        scheduledExecutorService.schedule({
                            instance.parent.destroy(instance.uuid)
                            instance.parent.destroy()

                            instance.message.thenAccept { message -> message.createUpdater().removeAllComponents().replaceMessage() }
                        }, 5, TimeUnit.MINUTES)
                    }
            }
        }
    }

    /**
     * Creates a description based on the available parameters of the feed model.
     *
     * @param server the server where this was executed.
     * @param feed   the feed to preview.
     * @param link   the value of the link to display.
     * @return a displayable description of the feed model.
     */
    private fun description(server: Server, feed: FeedModel, link: String): String {
        val subscribed = feed.mentions.joinToString(" ") {
                role -> server.getRoleById(role).map { it.mentionTag }.orElse("")
        }

        val channel = server.getTextChannelById(feed.channel).map { it.mentionTag }.orElse("Channel Not Found ❓")

        return ("ID: ${feed.unique}"
                + "\nLink: $link"
                + "\nName: ${feed.name}"
                + "\nRoles Subscribed: $subscribed"
                + "\nLast Update: `${feed.date}`"
                + "\nChannel: $channel"
                + "\nCreated by: <@${feed.user}>"
                + "\nStatus: ${accessible(feed.accessible)}")
    }

    private fun accessible(status: Boolean) = if (status) "**ONLINE**" else "**__ERROR__**"

    /**
     * Creates an embed based on the feed model and whether the user is authorized to see
     * the link (if the feed is of a reading list).
     *
     * @param server the server where this was executed.
     * @param feed   the feed model to preview.
     * @param user   the user who executed this.
     * @return an embed that contains all the information of the user.
     */
    private fun embed(server: Server, feed: FeedModel, user: User): EmbedBuilder {
        var link = feed.feedUrl + " [**(DO NOT LEAK)**](https://github.com/Amelia-chan/Amelia/discussions/19)"

        if (link.contains("unq=") && user.id != feed.user) {
            link = redactListLink(link)
        }

        return EmbedBuilder()
            .setTimestampToNow()
            .setColor(Color.YELLOW)
            .setTitle(feed.name)
            .setDescription(description(server, feed, link))
    }

    /**
     * Creates an embed based on the current cursor of the paginator.
     *
     * @param server the server where this was executed.
     * @param cursor the cursor from the paginator.
     * @return an embed that contains all the feeds available from the cursor.
     */
    private fun embed(server: Server, cursor: NexusPaginatorCursor<List<FeedModel>>): EmbedBuilder {
        val embed = EmbedBuilder().setTimestampToNow().setColor(Color.YELLOW).setTitle("${server.name}'s feeds")
            .setDescription("You are viewing **${cursor.displayablePosition}** out of **${cursor.maximumPages}** pages of feeds registered for this server.")
            .setFooter("Page ${cursor.displayablePosition} out of ${cursor.maximumPages}")

        for (feed in cursor.item) {
            var link = feed.feedUrl

            if (link.contains("unq=")) {
                link = redactListLink(link)
            }

            embed.addField("[${feed.unique}] ${feed.name}", description(server, feed, link))
        }

        return embed
    }


}